docs/compartmentalizing_dav1d.md

```diff
@@ -0,0 +1,577 @@
+and because the `fprintf` logging is in `libc` and shared anyways, this is fine.
+
+However, as the function pointer type is defined with `struct __va_list_tag *`,
+and because `__va_list_tag` is a private type, this will fail to compile.
+Instead, we can use `va_list` rather than `struct __va_list_tag *`.
+
+Fixing this properly in `ia2-rewriter` (by emitting just `va_list`)
+is tracked in [#429](https://github.com/immunant/IA2-Phase2/issues/429).
+
+#### `dlsym`
+
+`dav1d` calls `dlsym` at one point to conditionally call a function
+if it exists at runtime (specifically `__pthread_get_minstack`).
+We don't have access to this function,
+so it can't berewritten and a call gate generated for it,
+so we have the same situation as above with a half call gate.
+But this is again a `libc` function, so it's going to be in
+the shared compartment anyways, so this is not a problem.
+This means we can just create a `struct IA2_fnptr_*`
+from it without a double-sided call gate, i.e.,
+we can find the mangled (but stable) `IA2_fnptr_*` name
+for this function pointer type and then create an instance of it with:
+
+```c
+(struct IA2_fnptr__ZTSFmPK14pthread_attr_tE) { .ptr = dlsym(RTLD_DEFAULT, "__pthread_get_minstack") }
+```
+
+Fixing this properly in `ia2-rewriter`
+(by using a dedicated macro instead of having to look up the mangled type suffix)
+is tracked in [#425](https://github.com/immunant/IA2-Phase2/issues/425).
+
+## Compiling
+
+Now that we've rewritten `dav1d`, both automatically and manually,
+we can start compiling the compartmentalized version.
+In doing this, we need more build system changes, both in `meson` and outside of it,
+as well as more source changes to fix runtime compartment violations.
+
+### `meson` Changes
+
+There are a host of other flags and new generated call gate sources
+that must be built for a compartmentalized build,
+but that we don't need for the earlier build
+that was for generating a `compile_commands.json`.
+We can separate these in `meson` with an option in `meson_options.txt`,
+which we named `ia2_enable` (a `bool`).
+We also have an `ia2_path` option, which is what also turns on
+the ia2 args we added for the `compile_commands.json` earlier.
+
+We can divide these extra args into a few categories.
+First, there are args for each compile command
+and there are args for each link command.
+Then there are args that are for every command,
+or only for building `libdav1d.so` (`src/`, `lib.c`, or compartment 2)
+or `dav1d` (`tools/`, `dav1d.c`, or compartment 1).
+
+#### Global Args
+
+For global compile args, we add these:
+
+* `-include`, `join_paths(dav1d_src_root, 'callgate_wrapper.h')`:
+    Here, this `callgate_wrapper.h` is the header of the call gates.
+    It's named this specifically because we set our `--output-prefix` to this.
+* `-Werror=incompatible-pointer-types`:
+  To detect errors around function pointers and their wrapped call gates types.
+* `-DIA2_PERMISSIVE_MODE=1`:
```

Comment:
Please add a note here saying that this is specific to the compartmentalized dav1d demo. It's probably useful functionality to have in general since you found it useful as a flag, but I want to plan where to put that code (both in what source file and in what binary) before just opening up a PR that implements it.

Reply:
Sorry, I'm not sure what you mean exactly?  I'm trying to say here you can add this if you want to enable permissive mode.

Reply:
I mean that nothing in this repo checks for `-DIA2_PERMISSIVE_MODE`, only the build system/code in the dav1d repo does.

Reply:
Oh, yes, I know what you mean now.  That I added this for `dav1d` as a way of including `<permissive_mode.h>` or not.  Not that it's a part of `ia2` to begin with.  Will do.

Reply:
Fixed in bd9c866.